From d23a47764c1d3ff092dfae96a607239a21d6ba98 Mon Sep 17 00:00:00 2001
From: Franck LENORMAND <franck.lenormand@nxp.com>
Date: Fri, 28 Sep 2018 17:27:30 +0200
Subject: [PATCH 4763/5242] MLK-19656: crypto: dcp: Copy IV for CBC chaining

commit  0c8ce1dc75a8519114baeb6a829e46d5597e22a5 from
https://source.codeaurora.org/external/imx/linux-imx.git

Properly copy the IV for external chaining if we
are performing a CBC operation.

Signed-off-by: Franck LENORMAND <franck.lenormand@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/crypto/mxs-dcp.c |   12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/drivers/crypto/mxs-dcp.c b/drivers/crypto/mxs-dcp.c
index 92d5aab..d0dfadb 100644
--- a/drivers/crypto/mxs-dcp.c
+++ b/drivers/crypto/mxs-dcp.c
@@ -381,9 +381,15 @@ static int mxs_dcp_aes_block_crypt(struct crypto_async_request *arq)
 		if (limit_hit)
 			break;
 	}
-	if (last_out_len >= AES_BLOCK_SIZE) {
-		memcpy(req->info, out_buf+(last_out_len-AES_BLOCK_SIZE),
-		       AES_BLOCK_SIZE);
+
+	/* Copy the IV for CBC for chaining */
+	if (!rctx->ecb) {
+		if (rctx->enc)
+			memcpy(req->info, out_buf+(last_out_len-AES_BLOCK_SIZE),
+				AES_BLOCK_SIZE);
+		else
+			memcpy(req->info, in_buf+(last_out_len-AES_BLOCK_SIZE),
+				AES_BLOCK_SIZE);
 	}
 
 	return ret;
-- 
1.7.9.5

